File: java/src/test/java/com/alipay/easysdk/util/generic/ClientTest.java
package com.alipay.easysdk.util.generic;

import com.alipay.easysdk.TestAccount.Mini;
import com.alipay.easysdk.factory.Factory;
import com.alipay.easysdk.kernel.util.ResponseChecker;
import com.alipay.easysdk.util.generic.models.AlipayOpenApiGenericResponse;
import com.alipay.easysdk.util.generic.models.AlipayOpenApiGenericSDKResponse;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

import static org.hamcrest.CoreMatchers.*;
import static org.hamcrest.MatcherAssert.assertThat;

public class ClientTest {

    @Before
    public void setUp() {
        Factory.setOptions(Mini.CONFIG);
    }

    @Test
    public void testSDKExecute() throws Exception {
        Map<String, Object> bizParams = new HashMap<>();
        bizParams.put("subject", "iPhone6 16G");
        bizParams.put("out_trade_no", "f4833085-0c46-4bb0-8e5f-622a02a4cffc");
        bizParams.put("total_amount", "0.10");

        java.util.Map<String, String> textParams = new java.util.HashMap<>();

        AlipayOpenApiGenericSDKResponse response = Factory.Util.Generic().sdkExecute("alipay.trade.app.pay", textParams, bizParams);

        Assert.assertThat(ResponseChecker.success(response), is(true));
        Assert.assertThat(response.body, containsString("app_id=2019022663440152&biz_content=%7B%22"
                + "out_trade_no%22%3A%22f4833085-0c46-4bb0-8e5f-622a02a4cffc%22%2C%22"
                + "total_amount%22%3A%220.10%22%2C%22subject%22%3A%22iPhone6+16G%22%7D&"
                + "charset=UTF-8&format=json&method=alipay.trade.app.pay"
                + "&notify_url=https%3A%2F%2Fwww.test.com%2Fcallback&sign="));
    }

    @Test
    public void testFileExecute() throws Exception {
        Map<String, String> textParams = new HashMap<>();
        textParams.put("image_type", "png");
        textParams.put("image_name", "海底捞");
        textParams.put("image_pid", "22088021822217233");

        Map<String, String> fileParams = new HashMap<>();
        fileParams.put("image_content", "src/test/resources/fixture/sample.png");

        AlipayOpenApiGenericResponse response = Factory.Util.Generic().fileExecute("alipay.offline.material.image.upload", textParams, null, fileParams);

        assertThat(ResponseChecker.success(response), is(true));
        assertThat(response.code, is("10000"));
        assertThat(response.msg, is("Success"));
        assertThat(response.subCode, is(nullValue()));
        assertThat(response.subMsg, is(nullValue()));
        assertThat(response.httpBody, not(nullValue()));

    }


    @Test
    public void testExecuteWithoutAppAuthToken() throws Exception {
        String outTradeNo = UUID.randomUUID().toString();
        AlipayOpenApiGenericResponse response = Factory.Util.Generic().execute(
                "alipay.trade.create", null, getBizParams(outTradeNo));

        assertThat(ResponseChecker.success(response), is(true));
        assertThat(response.code, is("10000"));
        assertThat(response.msg, is("Success"));
        assertThat(response.subCode, is(nullValue()));
        assertThat(response.subMsg, is(nullValue()));
        assertThat(response.httpBody, not(nullValue()));
    }

    @Test
    public void testExecuteWithAppAuthToken() throws Exception {
        String outTradeNo = UUID.randomUUID().toString();
        AlipayOpenApiGenericResponse response = Factory.Util.Generic().execute(
                "alipay.trade.create", getTextParams(), getBizParams(outTradeNo));

        assertThat(ResponseChecker.success(response), is(false));
        assertThat(response.code, is("20001"));
        assertThat(response.msg, is("Insufficient Token Permissions"));
        assertThat(response.subCode, is("aop.invalid-app-auth-token"));
        assertThat(response.subMsg, is("无效的应用授权令牌"));
        assertThat(response.httpBody, not(nullValue()));
    }

    private Map<String, String> getTextParams() {
        Map<String, String> bizParams = new HashMap<>();
        bizParams.put("app_auth_token", "201712BB_D0804adb2e743078d1822d536956X34");
        return bizParams;
    }

    private Map<String, Object> getBizParams(String outTradeNo) {
        Map<String, Object> bizParams = new HashMap<>();
        bizParams.put("subject", "Iphone6 16G");
        bizParams.put("out_trade_no", outTradeNo);
        bizParams.put("total_amount", "0.10");
        bizParams.put("buyer_id", "2088002656718920");
        bizParams.put("extend_params", getHuabeiParams());
        return bizParams;
    }

    private Map<String, String> getHuabeiParams() {
        Map<String, String> extendParams = new HashMap<>();
        extendParams.put("hb_fq_num", "3");
        extendParams.put("hb_fq_seller_percent", "3");
        return extendParams;
    }
}

File: php/test/util/generic/ClientTest.php
<?php


namespace Alipay\EasySDK\Test\util\generic;


use Alipay\EasySDK\Kernel\Factory;
use Alipay\EasySDK\Test\TestAccount;
use Alipay\EasySDK\Kernel\Util\ResponseChecker;
use PHPUnit\Framework\TestCase;

class ClientTest extends TestCase
{
    public function __construct($name = null, array $data = [], $dataName = '')
    {
        parent::__construct($name, $data, $dataName);
        $account = new TestAccount();
        Factory::setOptions($account->getTestAccount());
    }

    public function testSDKExecute()
    {
        $bizParams = array(
            "subject" => "Iphone6 16G",
            "out_trade_no" => "f4833085-0c46-4bb0-8e5f-622a02a4cffc",
            "total_amount" => "0.10"
        );

        $textParams = array();

        $result = Factory::util()->generic()->sdkExecute("alipay.trade.app.pay", $textParams, $bizParams);
        $this->assertEquals(true, strpos($result->body, 'alipay_sdk=alipay-easysdk-php') > 0);
        $this->assertEquals(true, strpos($result->body, 'sign') > 0);
    }

    public function testFileExecute()
    {
        $textParams = array(
            "image_type" => "png",
            "image_name" => "海底捞",
            "image_pids" => "22088021822217233"
        );
        $account = new TestAccount();
        $filePath = $account->getResourcesPath(). '/resources/fixture/sample.png';

        $fileParams = array(
            "image_content" => $filePath
        );

        $result = Factory::util()->generic()->fileExecute("alipay.offline.material.image.upload", $textParams, null, $fileParams);
        $responseChecker = new ResponseChecker();
        $this->assertEquals(true, $responseChecker->success($result));
        $this->assertEquals('Success', $result->msg);
    }

    public function testExecuteWithoutAppAuthToken()
    {
        $result = Factory::util()->generic()->execute("alipay.trade.create", null, $this->getBizParams(microtime()));
        $this->assertEquals('10000', $result->code);
        $this->assertEquals('Success', $result->msg);
    }

    public function testExecuteWithAppAuthToken()
    {
        $result = Factory::util()->generic()->execute("alipay.trade.create", $this->getTextParams(), $this->getBizParams(microtime()));
        $this->assertEquals('20001', $result->code);
        $this->assertEquals('Insufficient Token Permissions', $result->msg);
        $this->assertEquals('aop.invalid-app-auth-token', $result->subCode);
        $this->assertEquals('无效的应用授权令牌', $result->subMsg);
    }

    //设置系统参数（OpenAPI中非biz_content里的参数）
    private function getTextParams()
    {
        return array("app_auth_token" => "201712BB_D0804adb2e743078d1822d536956X34");
    }

    private function getBizParams($outTradeNo)
    {
        $bizParams = array(
            "subject" => "Iphone6 16G",
            "out_trade_no" => $outTradeNo,
            "total_amount" => "0.10",
            "buyer_id" => "2088002656718920",
            "extend_params" => $this->getHuabeiParams()
        );
        return $bizParams;
    }

    private function getHuabeiParams()
    {
        $extendParams = array("hb_fq_num" => "3", "hb_fq_seller_percent" => "3");
        return $extendParams;
    }
}